### ISSUE TITLE
Make a rule for variables declared without an explicit type

### ISSUE THREAD

**Issue by nre-ableton** [CodeNarc/CodeNarc#469]
So all of these variable declarations are valid in groovy:

```groovy
def foo = 'bar'  // Bad, but CodeNarc will catch this with NoDef
String foo = 'bar'  // Ideal
foo = 'bar'  // Bad, but CodeNarc won't catch this
```

That last case should be treated as a violation. I'm not sure whether it makes sense to extend the `NoDef` rule to catch this, or if a new rule should be introduced. I'd be fine with either solution.

**Comment by chrismair** [comment]
I don't think that last case is even legal Groovy: `foo = 'bar'`, at least not by itself.

**Comment by nre-ableton** [comment]
Regardless of whether it is strictly legal, it works in Jenkins pipeline scripts, which use Groovy 2.4.12.

**Comment by erdi** [comment]
If you are doing this in a script then this is an assignment to a variable
in the script's binding and not a variable declaration.

**Comment by nre-ableton** [comment]
@erdi This also works, though:

```
/tmp$> cat hi.groovy
#!/usr/bin/env groovy
foo = 'bar'
println foo
/tmp$> groovy hi.groovy
bar
```

🤷‍♂

**Comment by erdi** [comment]
I'm not saying that it does not work. What I'm saying is that it's not a
declaration but an assignment.

**Comment by nre-ableton** [comment]
@erdi Ok, that's a valid point. Nonetheless, it would be nice if CodeNarc could warn about this. 😄 I'll edit the issue title/description though.

**Comment by chrismair** [comment]
Okay, given that as soon as you specify a type for `foo`, it becomes a *local variable* within the `run` method, rather than a script binding, I guess what is being requested here is a rule to flag/forbid *all* script bindings.

**Comment by nre-ableton** [comment]
@chrismair Thanks for the clarification! I was not entirely certain what was happening under the hood here.
